axelor-base/src/main/resources/views/PricingRule.xml

```diff
@@ -26,13 +27,30 @@
       <field name="operatorSelect" showIf="typeSelect == 1 &amp;&amp; fieldTypeSelect != 0"/>
       <field name="formula" widget="CodeEditor" x-code-syntax="groovy" colSpan="12"/>
       <field name="fieldToPopulate" domain="self.metaModel = :concernedModel"
-        showIf="typeSelect == 2"/>
+        showIf="typeSelect == 2" onChange="action-pricing-rule-attrs-display-meta-json-field"/>
       <field name="tempVarName" showIf="typeSelect == 2"/>
+      <field name="metaJsonField" title="Custom field"
+        onSelect="action-pricing-rule-attrs-on-select-meta-json-field" hidden="true"/>
     </panel>
   </form>
 
   <action-attrs name="action-pricing-rule-attrs-field-to-populate-clear">
     <attribute name="value" for="fieldToPopulate" expr="eval: null"/>
   </action-attrs>
 
+  <action-attrs name="action-pricing-rule-attrs-on-select-meta-json-field">
+    <attribute name="domain" for="metaJsonField"
+      expr="eval:&quot; self.type NOT IN ('panel', 'button', 'label', 'spacer', 'separator') AND self.model = '${concernedModel?.fullName}' AND self.modelField = '${fieldToPopulate?.name}' &quot;"/>
+  </action-attrs>
+
+  <action-attrs name="action-pricing-rule-attrs-display-meta-json-field">
+    <attribute name="hidden" for="metaJsonField"
+      expr="eval:__this__?.fieldToPopulate.json == false"
+      if="eval:__this__?.fieldToPopulate != null"/>
+    <attribute name="required" for="metaJsonField"
+      expr="eval:__this__?.fieldToPopulate.json == true" if="eval:__this__?.fieldToPopulate != null"/>
+    <attribute name="value" for="metaJsonField" expr="eval:null"
+      if="eval:__this__?.fieldToPopulate== null || __this__?.fieldToPopulate.json == false"/>
```

Comment:
be careful when using safe dereferencing with boolean comparison since `null == false` will return `false`. But I don't know if this cause an issue since I suspect that `__this__` will never be null